### PR TITLE
feat(ingestion/glue): delta schemas

### PR DESCRIPTION
**Pull request by sgomezvillamor**
Hive sync up in Spark is wrongly reported as `col (array<string>)`. Experience shows that the issue happens only when using Scala and for tables (not views), which is actually a very common case. Given that the correct schema is serialized in the `spark.sql.sources.schema.part.{i}` table parameters, this PR parses and processes the schema from those properties so such a valuable metadata is not lost.

This is solving the following feature request: https://feature-requests.datahubproject.io/p/glue-crawler-support-for-correct-delta-schema-stored-in-properties

This bug in Spark has been reported many times and never solved. Some references here:
* https://github.com/delta-io/delta/issues/1679
* https://github.com/delta-io/delta/pull/2310

## Checklist

- [ ] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [ ] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable). If a new feature has been added a Usage Guide has been added for the same.
- [ ] For any breaking change/potential downtime/deprecation/big changes an entry has been made in [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md)

### CODE REVIEW

File: metadata-ingestion/tests/integration/delta_lake/golden_files/local/golden_mces_tables_with_nested_datatypes.json [hunk 1]
```diff
@@ -96,11 +96,11 @@
                                 "jsonProps": "{\"native_data_type\": \"struct<_1:long,_2:string>\"}"
                             },
                             {
-                                "fieldPath": "[version=2.0].[type=struct].[type=struct].data.[type=struct]._2.[type=null]._1",
+                                "fieldPath": "[version=2.0].[type=struct].[type=struct].data.[type=struct]._2.[type=long]._1",
                                 "nullable": true,
                                 "type": {
                                     "type": {
-                                        "com.linkedin.pegasus2avro.schema.NullType": {}
+                                        "com.linkedin.pegasus2avro.schema.NumberType": {}
```
Comment (sgomezvillamor) [inline metadata-ingestion/tests/integration/delta_lake/golden_files/local/golden_mces_tables_with_nested_datatypes.json:103]: Note for reviewers:
This is for the new mapping I added here 
https://github.com/datahub-project/datahub/blob/21b3c4773f3aac54767cd5291d6d249bc88a1346/metadata-ingestion/src/datahub/utilities/hive_schema_to_avro.py#L31

File: metadata-ingestion/tests/integration/delta_lake/golden_files/local/golden_mces_tables_with_nested_datatypes.json [hunk 2]
```diff
@@ -1796,397 +1796,6 @@
         "lastRunId": "no-run-id-provided"
     }
 },
-{
-    "entityType": "dataset",
-    "entityUrn": "urn:li:dataset:(urn:li:dataPlatform:delta-lake,my-platform.tests/integration/delta_lake/test_data/delta_tables_nested_datatype/table_with_string_and_nested_array_of_numbers,UAT)",
-    "changeType": "UPSERT",
-    "aspectName": "browsePathsV2",
```
Comment (sgomezvillamor) [inline metadata-ingestion/tests/integration/delta_lake/golden_files/local/golden_mces_tables_with_nested_datatypes.json:1803]: Note for reviewers:
No idea why these `browsePathsV2` aspects are gone, any clue?

File: metadata-ingestion/src/datahub/ingestion/source/aws/glue.py [hunk 3]
```diff
@@ -204,6 +206,8 @@ class GlueSourceReport(StaleEntityRemovalSourceReport):
     num_job_script_failed_parsing: int = 0
     num_job_without_nodes: int = 0
     num_dataset_to_dataset_edges_in_job: int = 0
+    num_dataset_schema_invalid: int = 0
+    num_dataset_buggy_delta_schema: int = 0
```
Comment (sgomezvillamor) [inline metadata-ingestion/src/datahub/ingestion/source/aws/glue.py:210]: Note for reviewers:
I added these two mainly for the testing. I'm ok to rename or remove even

File: metadata-ingestion/tests/integration/delta_lake/golden_files/local/golden_mces_tables_with_nested_datatypes.json (hunk 2 above)
Comment (sgomezvillamor) [reply metadata-ingestion/tests/integration/delta_lake/golden_files/local/golden_mces_tables_with_nested_datatypes.json:1803]: Definitely, this is related to https://github.com/datahub-project/datahub/blob/21b3c4773f3aac54767cd5291d6d249bc88a1346/metadata-ingestion/src/datahub/utilities/hive_schema_to_avro.py#L31

As soon as I revert that, `browsePathsV2` aspects are back. So this requires some investigation

Comment (sgomezvillamor) [reply metadata-ingestion/tests/integration/delta_lake/golden_files/local/golden_mces_tables_with_nested_datatypes.json:1803]: What I have found is that there are no missing `browsePathsV2` aspects actually. They were duplicated in the golden file and not anymore!

File: metadata-ingestion/tests/integration/delta_lake/golden_files/local/golden_mces_tables_with_nested_datatypes.json [hunk 4]
```diff
@@ -1796,397 +1796,6 @@
         "lastRunId": "no-run-id-provided"
     }
 },
-{
```
Comment (sgomezvillamor) [inline metadata-ingestion/tests/integration/delta_lake/golden_files/local/golden_mces_tables_with_nested_datatypes.json:1799]: Note for the reviewers:

Removed aspects are not _lost_; they were duplicated and updating the golden file resulted on removing the duplicated aspects.

**Comment by sgomezvillamor** [comment]
@hsheth2 or @treff7es , as usual contributors to ingestion codebase, could you have a look at this?

File: metadata-ingestion/src/datahub/ingestion/source/aws/glue.py (hunk 3 above)
Comment (mayurinehate) [reply metadata-ingestion/src/datahub/ingestion/source/aws/glue.py:210]: This is nice. Let's please rename as

num_dataset_schema_invalid -> num_dataset_invalid_delta_schema
num_dataset_buggy_delta_schema -> num_dataset_valid_delta_schema

File: metadata-ingestion/src/datahub/ingestion/source/aws/glue.py [hunk 5]
```diff
@@ -1148,9 +1152,35 @@ def get_s3_tags() -> Optional[GlobalTagsClass]:
             return new_tags
 
         def get_schema_metadata() -> Optional[SchemaMetadata]:
-            if not table.get("StorageDescriptor"):
+            def is_delta_schema(columns: Optional[List[Mapping[str, Any]]]) -> bool:
```
Comment (mayurinehate) [inline metadata-ingestion/src/datahub/ingestion/source/aws/glue.py:1155]: Defining functions within other functions is discouraged as per [coding style](https://datahubproject.io/docs/metadata-ingestion/developing/#code-style). Can you please move this function outside ?

Comment (mayurinehate) [reply metadata-ingestion/src/datahub/ingestion/source/aws/glue.py:1155]: Can this be refractored to accept both tableParameters and tableStorageDescriptor and then return boolean ? this function can subsume this check as well -> `(provider == "delta") and (num_parts > 0)`

File: metadata-ingestion/src/datahub/ingestion/source/aws/glue.py [hunk 6]
```diff
@@ -1148,9 +1152,35 @@ def get_s3_tags() -> Optional[GlobalTagsClass]:
...
+                return _get_delta_schema_metadata()
+
+            elif table.get("StorageDescriptor"):
+                return _get_glue_schema_metadata()
+
+            else:
                 return None
 
+        def _get_glue_schema_metadata() -> Optional[SchemaMetadata]:
+            assert table.get("StorageDescriptor")
```
Comment (mayurinehate) [inline metadata-ingestion/src/datahub/ingestion/source/aws/glue.py:1182]: can remove this assert as this is already checked earlier.

**Review by mayurinehate — COMMENTED** [review]
Can we add config `extract_delta_schema_from_parameters` to enable this behavior. can be set to default True. If disabled, none of this delta customization would be used.

Can you confirm if this issue is with particular spark/delta version or all of them ?

Also, if I understand correctly, this workaround can entirely be removed if the issue is fixed in new spark/delta version - for example - if the delta.io PR that you have linked is merged, right ? If so, please add a comment in codebase to mention this.

File: metadata-ingestion/src/datahub/ingestion/source/aws/glue.py (hunk 5 above)
Comment (sgomezvillamor) [reply metadata-ingestion/src/datahub/ingestion/source/aws/glue.py:1155]: > Defining functions within other functions is discouraged as per [coding style](https://datahubproject.io/docs/metadata-ingestion/developing/#code-style). Can you please move this function outside ?

While I do agree, I just followed the existing pattern in the code. Note `get_owner`, `get_dataset_properties`, `get_s3_tags` are all defined within `_extract_record` method.

So, are you suggesting to move `is_delta_schema` at `_extract_record` level too or at `GlueSource` level?

**Comment by sgomezvillamor** [comment]
Thanks for the review @mayurinehate . I will address comments soon.

> Can we add config `extract_delta_schema_from_parameters` to enable this behavior. can be set to default True. If disabled, none of this delta customization would be used.

Sure, good point!

> Can you confirm if this issue is with particular spark/delta version or all of them ?

Not a particular version but an long-time overall issue. As you can see in the two links listed in the PR description, community is recurrently complaining about this issue. 

> Also, if I understand correctly, this workaround can entirely be removed if the issue is fixed in new spark/delta version - for example - if the delta.io PR that you have linked is merged, right ? If so, please add a comment in codebase to mention this.

True. As soon as the hive integration with Spark is correctly providing the schema as expected in the `StorageProperties`, this could be removed. If that will happen in the PR I linked or some other, hard to predict 😄 
I will provide more context in the comment here https://github.com/datahub-project/datahub/blob/dbceda1d5f1678c1438c4848497cb3d48356fc44/metadata-ingestion/src/datahub/ingestion/source/aws/glue.py#L1163

File: metadata-ingestion/src/datahub/ingestion/source/aws/glue.py (hunk 3 above)
Comment (sgomezvillamor) [reply metadata-ingestion/src/datahub/ingestion/source/aws/glue.py]: addressed in https://github.com/datahub-project/datahub/pull/10299/commits/d634b10870bd0408cf3e5497c235bb5f0c4070a8

File: metadata-ingestion/src/datahub/ingestion/source/aws/glue.py (hunk 5 above)
Comment (sgomezvillamor) [reply metadata-ingestion/src/datahub/ingestion/source/aws/glue.py]: As just moved the `is_delta_schema` at `_extra_record` level, to keep it aligned with all others existing methods in the codebase.

**Comment by sgomezvillamor** [comment]
Added the `extract_delta_schema_from_parameters ` config option with `False` default value in https://github.com/datahub-project/datahub/pull/10299/commits/d634b10870bd0408cf3e5497c235bb5f0c4070a8

**Comment by sgomezvillamor** [comment]
@mayurinehate Could you please review again? Thanks

File: metadata-ingestion/src/datahub/ingestion/source/aws/glue.py [hunk 7]
```diff
@@ -163,6 +163,11 @@ class GlueSourceConfig(
     stateful_ingestion: Optional[StatefulStaleMetadataRemovalConfig] = Field(
         default=None, description=""
     )
+    extract_delta_schema_from_parameters: Optional[bool] = Field(
+        default=False,
+        description="If enabled, delta schemas can be alternatively fetched from table parameters "
+        "(https://github.com/delta-io/delta/pull/2310)",
```
Comment (mayurinehate) [inline metadata-ingestion/src/datahub/ingestion/source/aws/glue.py:169]:
```suggestion
    extract_delta_schema_from_parameters: Optional[bool] = Field(
        default=False,
        description="If enabled, delta schemas can be alternatively fetched from table parameters."
```

The PR is already present in code comment later so need not be repeated.

**Review by mayurinehate — APPROVED** [review]
Minor edits suggested. Everything else looks good to me.

**Comment by sgomezvillamor** [comment]
@mayurinehate Thanks for the feedback and approval.
I have added your latests suggestions and solved conflicts.

Build is all green but a nifi test. Surprisingly the test complains about not matching golden fine in 3.10 while it works fine in 3.8. Do you think is that related to my updates? or just some _random_ error?

**Comment by mayurinehate** [comment]
CI Failures are unrelated to the changes in the PR.

**Review by treff7es — APPROVED** [review]

**Pull request merged**